packages/realm/src/Counter.ts

```diff
@@ -0,0 +1,140 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2024 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+import { IllegalConstructorError, Realm, UpdateMode, assert, binding } from "./internal";
+
+const REALM = Symbol("Counter#realm");
+const OBJ = Symbol("Counter#obj");
+const COLUMN_KEY = Symbol("Counter#columnKey");
+
+/**
+ * A logical counter representation for performing numeric updates that need
+ * to be synchronized as sequentially consistent events rather than individual
+ * reassignments of the number.
+ *
+ * For instance, offline Client 1 and Client 2 which both see `Counter.value`
+ * as `0`, can both call `Counter.increment(1)`. Once online, the value will
+ * converge to `2`.
+ * @note
+ * __Declaring a counter__:
+ *
+ * A property schema is declared as either:
+ * - `"counter"`
+ * - `{ type: "int", presentation: "counter" }`
+ *
+ * __Creating a counter__:
+ *
+ * Use a `number` when creating your counter on a {@link Realm.Object}.
+ * ```ts
+ * realm.write(() => {
+ *   realm.create(MyObject, { _id, counter: 0 });
+ * })
+ * ```
+ *
+ * To update a `null` counter, use {@link UpdateMode.Modified} or {@link UpdateMode.All}.
+ * ```ts
+ * realm.write(() => {
+ *   realm.create(MyObject, { _id, counter: 0 }, UpdateMode.Modified);
+ * })
+ * ```
+ *
+ * __Nullability__:
+ *
+ * The above property schema can be extended to allow a nullable counter.
+ * A `Counter` never stores `null` values itself, but the counter property
+ * on the {@link Realm.Object} (e.g. `myRealmObject.myCounter`) can be `null`.
+ *
+ * __Not a live object__:
+ *
+ * The `Counter` instance itself is _not_ a live object. Calling `myCounter.value`
+ * will always return the most recent locally persisted (and synced) state, and
+ * calling `myRealmObject.myCounter` always returns a new `Counter` instance (or
+ * potentially `null` for nullable counters).
+ */
+export class Counter {
+  /** @internal */
+  private readonly [REALM]: Realm;
+  /** @internal */
+  private readonly [OBJ]: binding.Obj;
+  /** @internal */
+  private readonly [COLUMN_KEY]: binding.ColKey;
+
+  /** @internal */
+  constructor(realm: Realm, obj: binding.Obj, columnKey: binding.ColKey) {
```

Comment:
P.S. Despite `@internal`, the constructors still show up in the API ref docs (for other classes as well).

Reply:
This is a bit weird 🤔 Perhaps there's something misconfigured in the way we generate docs then.